### PR TITLE
Distributed Trainer

### PR DESCRIPTION
**Pull request by arunppsg**
DistributedTrainer for DeepChem.

### CODE REVIEW

File: deepchem/models/lightning/dc_lightning_dataset_module.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ class DCLightningDatasetModule(pl.LightningDataModule):
     This class requires PyTorch to be installed.
     """
 
-    def __init__(self, dataset, batch_size, collate_fn):
+    def __init__(self, dataset, batch_size, collate_fn, num_workers: int):
```
Comment (rbharath) [inline deepchem/models/lightning/dc_lightning_dataset_module.py:29]: docs reminder

Comment (rbharath) [reply deepchem/models/lightning/dc_lightning_dataset_module.py:29]: Let's just add type annotations while we're at it here

File: deepchem/models/trainer.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import deepchem as dc
+from deepchem.models.lightning.dc_lightning_module import DCLightningModule
+from deepchem.models.lightning.dc_lightning_dataset_module import DCLightningDatasetModule, collate_dataset_wrapper
+
+class DistributedTrainer():
```
Comment (rbharath) [inline deepchem/models/trainer.py:5]: Let's add more info here about the algorithm, how it works under the hood. Make at least a page of docs. Add to rst

Comment (arunppsg) [reply deepchem/models/trainer.py]: Added docs and to rst

File: deepchem/models/lightning/dc_lightning_dataset_module.py (hunk 1 above)
Comment (arunppsg) [reply deepchem/models/lightning/dc_lightning_dataset_module.py]: Added type annotations and docs

**Review by rbharath — COMMENTED** [review]
For now, let's not make trainer a top level module but let's add `dc.models.trainer` as a submodule. Once those changes are made and tests are fixed, we should be good to merge in

**Review by rbharath — APPROVED** [review]
LGTM, but there's a bug that needs to be fixed before merge on the CI for the distributed trainer.

**Comment by arunppsg** [comment]
Fixed CI error and merging this in.

**Pull request merged**